### PR TITLE
Added interdomain setup and testing

### PR DESCRIPTION
**Pull request by Mixaster995**
Signed-off-by: Mikhail Avramenko <avramenkomihail15@gmail.com>

## Issue
closes https://github.com/networkservicemesh/integration-k8s-kind/issues/517

### CODE REVIEW

File: entry_point_test.go [hunk 1]
```diff
@@ -17,28 +17,49 @@
...
+	if os.Getenv("INTERDOMAIN-CI") == "enabled" {
+		t.Skip()
+	}
 	suite.Run(t, new(heal.Suite))
 }
 
 func TestRunFeatureSuite(t *testing.T) {
+	if os.Getenv("INTERDOMAIN-CI") == "enabled" {
+		t.Skip()
+	}
```
Comment (denis-tingaikin) [inline entry_point_test.go:42]: ## Option 1
Consider using build tags.

Note: you can separate tests into two files with two tags
```go
// file1.go

// +build interdomain
```
...
```go
// file2. go

// +build single
```
And running of test

```go
go test -tags=interdomain
go test -tags=single
```

See at https://pkg.go.dev/go/build.


## Option 2

Run tests via Match (where names of test should contain "Interdomain" or "Single", for example, "Tes1Single", "Test2Interdomain")
  
```
go test -run Intedomain
go test -run Signle
```

Or via pkg
```
go test pkg/single
go test pkg/interdomain
```

Comment (Mixaster995) [reply entry_point_test.go]: ok, done
used build tags

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -116,3 +116,78 @@ jobs:
...
+          echo $GITHUB_WORKSPACE/bin >> $GITHUB_PATH
+      - uses: actions/checkout@v2
+        with:
+          path: ${{ github.workspace }}/src/github.com/${{ github.repository }}
+      - name: Get kind
+        run: go get sigs.k8s.io/kind@v0.11.1
+      - name: Create kind clusters
+        env:
+          CONFIG: 'cluster-config-interdomain.yaml'
+          TIMEOUT: 200s
```
Comment (denis-tingaikin) [inline .github/workflows/ci.yaml:145]:
```suggestion
```

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -116,3 +116,78 @@ jobs:
...
+          echo GO111MODULE=on >> $GITHUB_ENV
+          echo $GITHUB_WORKSPACE/bin >> $GITHUB_PATH
+      - uses: actions/checkout@v2
+        with:
+          path: ${{ github.workspace }}/src/github.com/${{ github.repository }}
+      - name: Get kind
+        run: go get sigs.k8s.io/kind@v0.11.1
+      - name: Create kind clusters
+        env:
+          CONFIG: 'cluster-config-interdomain.yaml'
```
Comment (denis-tingaikin) [inline .github/workflows/ci.yaml:144]:
```suggestion
```

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -116,3 +116,78 @@ jobs:
...
+          TIMEOUT: 200s
+          CLUSTER_NAME_1: interdomain-kind-1
+          CLUSTER_NAME_2: interdomain-kind-2
+          CLUSTER_NAME_3: interdomain-kind-3
+          KUBE_CONFIG_PATH_PREFIX: ${{ github.workspace }}/src/github.com/${{ github.repository }}
+        run: |
+          IMAGE="kindest/node:$KUBERNETES_VERSION"
+          if [[ $KUBERNETES_VERSION=="" ]]; then
+              IMAGE="kindest/node:v1.22.1"
+          fi
```
Comment (denis-tingaikin) [inline .github/workflows/ci.yaml:154]:
```suggestion
```

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -116,3 +116,78 @@ jobs:
...
+      - name: Create kind clusters
+        env:
+          CONFIG: 'cluster-config-interdomain.yaml'
+          TIMEOUT: 200s
+          CLUSTER_NAME_1: interdomain-kind-1
+          CLUSTER_NAME_2: interdomain-kind-2
+          CLUSTER_NAME_3: interdomain-kind-3
+          KUBE_CONFIG_PATH_PREFIX: ${{ github.workspace }}/src/github.com/${{ github.repository }}
+        run: |
+          IMAGE="kindest/node:$KUBERNETES_VERSION"
```
Comment (denis-tingaikin) [inline .github/workflows/ci.yaml:151]:
```suggestion
```

File: .github/workflows/ci.yaml [hunk 6]
```diff
@@ -116,3 +116,78 @@ jobs:
...
+          path: ${{ github.workspace }}/src/github.com/${{ github.repository }}
+      - name: Get kind
+        run: go get sigs.k8s.io/kind@v0.11.1
+      - name: Create kind clusters
+        env:
+          CONFIG: 'cluster-config-interdomain.yaml'
+          TIMEOUT: 200s
+          CLUSTER_NAME_1: interdomain-kind-1
+          CLUSTER_NAME_2: interdomain-kind-2
+          CLUSTER_NAME_3: interdomain-kind-3
```
Comment (denis-tingaikin) [inline .github/workflows/ci.yaml:148]:
```suggestion
```

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -116,3 +116,78 @@ jobs:
...
+        run: |
+          IMAGE="kindest/node:$KUBERNETES_VERSION"
+          if [[ $KUBERNETES_VERSION=="" ]]; then
+              IMAGE="kindest/node:v1.22.1"
+          fi
+
+          echo "creating kind clusters"
+          kind create cluster --name $CLUSTER_NAME_1 --config $CONFIG --wait $TIMEOUT --image="$IMAGE"
+          kind create cluster --name $CLUSTER_NAME_2 --config $CONFIG --wait $TIMEOUT --image="$IMAGE"
+          kind create cluster --name $CLUSTER_NAME_3 --config $CONFIG --wait $TIMEOUT --image="$IMAGE"
```
Comment (denis-tingaikin) [inline .github/workflows/ci.yaml:159]:
```suggestion
          for i in {1..3} do
          kind create cluster --name kind-${i} --config cluster-config-interdomain.yaml --image="kindest/node:$NSM_KUBERNETES_VERSION"
          done
```

File: .github/workflows/ci.yaml [hunk 8]
```diff
@@ -116,3 +116,78 @@ jobs:
...
+          echo "config 3:"
+          cat $KUBE_CONFIG_PATH_PREFIX/config3
+
+          echo KUBECONFIG1="$KUBE_CONFIG_PATH_PREFIX/config1" >> $GITHUB_ENV
+          echo KUBECONFIG2="$KUBE_CONFIG_PATH_PREFIX/config2" >> $GITHUB_ENV
+          echo KUBECONFIG3="$KUBE_CONFIG_PATH_PREFIX/config3" >> $GITHUB_ENV
+
+          echo CLUSTER1_CIDR="172.18.1.128/25" >> $GITHUB_ENV
+          echo CLUSTER2_CIDR="172.18.2.128/25" >> $GITHUB_ENV
+          echo CLUSTER3_CIDR="172.18.3.128/25" >> $GITHUB_ENV
```
Comment (denis-tingaikin) [inline .github/workflows/ci.yaml:180]:
```suggestion
          for i in {1..3} do
          configPath=${{ github.workspace }}/src/github.com/${{ github.repository }}/config${i}
    kind get kubeconfig --name "kind-${i}" > ${configPath}
          echo KUBECONFIG${i}=${configPath}
          echo CLUSTER${i}_CIDR="172.18.${i}.128/25" >> $GITHUB_ENV
done
```

File: .github/workflows/ci.yaml [hunk 9]
```diff
@@ -116,3 +116,78 @@ jobs:
...
+        working-directory: ${{ github.workspace }}/src/github.com/${{ github.repository }}
+      - name: Interdomain tests
+        run: |
+          go test -count 1 -timeout 1h -race -v -run "[a-zA-Z]*Interdomain[a-zA-Z]*"
+        env:
+          ARTIFACTS_DIR: interdomain-logs
+        working-directory: ${{ github.workspace }}/src/github.com/${{ github.repository }}
+      - name: Upload artifacts
+        if: ${{ success() || failure() || cancelled() }}
```
Comment (denis-tingaikin) [inline .github/workflows/ci.yaml:189]:
```suggestion
      - name: Cleanup resources
        if: ${{ success() || failure() || cancelled() }}
        run: kind delete clusters $(kind get clusters)
      - name: Upload artifacts
        if: ${{ success() || failure() || cancelled() }}
```

File: .github/workflows/ci.yaml [hunk 10]
```diff
@@ -106,7 +106,7 @@ jobs:
         working-directory: ${{ github.workspace }}/src/github.com/${{ github.repository }}
       - name: Integration tests
         run: |
-          go test -count 1 -timeout 1h -race -v ./...
+          go test -count 1 -timeout 1h -race -v -run "[a-zA-Z]*Single[a-zA-Z]*"
```
Comment (denis-tingaikin) [inline .github/workflows/ci.yaml:109]:
```suggestion
          go test -count 1 -timeout 1h -race -v -run Signle
```

File: .github/workflows/ci.yaml [hunk 11]
```diff
@@ -116,3 +116,78 @@ jobs:
...
+          echo KUBECONFIG2="$KUBE_CONFIG_PATH_PREFIX/config2" >> $GITHUB_ENV
+          echo KUBECONFIG3="$KUBE_CONFIG_PATH_PREFIX/config3" >> $GITHUB_ENV
+
+          echo CLUSTER1_CIDR="172.18.1.128/25" >> $GITHUB_ENV
+          echo CLUSTER2_CIDR="172.18.2.128/25" >> $GITHUB_ENV
+          echo CLUSTER3_CIDR="172.18.3.128/25" >> $GITHUB_ENV
+        working-directory: ${{ github.workspace }}/src/github.com/${{ github.repository }}
+      - name: Interdomain tests
+        run: |
+          go test -count 1 -timeout 1h -race -v -run "[a-zA-Z]*Interdomain[a-zA-Z]*"
```
Comment (denis-tingaikin) [inline .github/workflows/ci.yaml:184]:
```suggestion
          go test -count 1 -timeout 1h -race -v -run Interdomain
```

File: README.md [hunk 12]
```diff
@@ -1,7 +1,41 @@
...
+```bash
+go test -count 1 -timeout 1h -race -v -run "[a-zA-Z]*Single[a-zA-Z]*"
+```
+
+## Multiple cluster scenario(interdomain tests)
+1. Create 3 kind clusters:
+```bash
+kind create cluster --name interdomain-kind-1 --config cluster-config-interdomain.yaml --wait 120s
+kind create cluster --name interdomain-kind-2 --config cluster-config-interdomain.yaml --wait 120s
+kind create cluster --name interdomain-kind-3 --config cluster-config-interdomain.yaml --wait 120s
```
Comment (denis-tingaikin) [inline README.md:22]:
```suggestion
kind create cluster --name kind-1 --config cluster-config-interdomain.yaml --wait 120s
kind create cluster --name kind-2 --config cluster-config-interdomain.yaml --wait 120s
kind create cluster --name kind-3 --config cluster-config-interdomain.yaml --wait 120s
```

File: README.md [hunk 13]
```diff
@@ -1,7 +1,41 @@
...
+
+1. Create kind cluster:
 ```bash
 kind create cluster --config cluster-config.yaml --wait 120s
-go test ./...
+```
+
+2. Run tests
+```bash
+go test -count 1 -timeout 1h -race -v -run "[a-zA-Z]*Single[a-zA-Z]*"
```
Comment (denis-tingaikin) [inline README.md:14]:
```suggestion
go test -count 1 -timeout 1h -race -v -run Single
```

File: README.md [hunk 14]
```diff
@@ -1,7 +1,41 @@
...
+kind create cluster --name interdomain-kind-1 --config cluster-config-interdomain.yaml --wait 120s
+kind create cluster --name interdomain-kind-2 --config cluster-config-interdomain.yaml --wait 120s
+kind create cluster --name interdomain-kind-3 --config cluster-config-interdomain.yaml --wait 120s
+```
+
+2. Save kubeconfig of each cluster(you may choose appropriate location)
+```bash
+kind get kubeconfig --name interdomain-kind-1 > /tmp/config1
+kind get kubeconfig --name interdomain-kind-2 > /tmp/config2
+kind get kubeconfig --name interdomain-kind-3 > /tmp/config3
```
Comment (denis-tingaikin) [inline README.md:29]:
```suggestion
kind get kubeconfig --name kind-1 > /tmp/config1
kind get kubeconfig --name kind-2 > /tmp/config2
kind get kubeconfig --name kind-3 > /tmp/config3
```

File: README.md [hunk 15]
```diff
@@ -1,7 +1,41 @@
...
+
+3. Run interdomain tests with necessary environment variables set
+```bash
+KUBECONFIG1==/tmp/config1 \
+KUBECONFIG2==/tmp/config2 \
+KUBECONFIG3==/tmp/config3 \
+CLUSTER1_CIDR="172.18.1.128/25" \
+CLUSTER2_CIDR="172.18.2.128/25"\
+CLUSTER3_CIDR="172.18.3.128/25"\
+bash -c 'go test -count 1 -timeout 1h -race -v -run "[a-zA-Z]*Interdomain[a-zA-Z]*"'
```
Comment (denis-tingaikin) [inline README.md:40]:
```suggestion
export KUBECONFIG1=/tmp/config1
export KUBECONFIG2=/tmp/config2
export KUBECONFIG3=/tmp/config3
export CLUSTER1_CIDR="172.18.1.128/25"
export CLUSTER2_CIDR="172.18.2.128/25"
export CLUSTER3_CIDR="172.18.3.128/25"
go test -count 1 -timeout 1h -race -v -run Interdomain
```

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]
Looks overcomplicated.

File: .github/workflows/ci.yaml (hunk 10 above)
Comment (Mixaster995) [reply .github/workflows/ci.yaml:109]: It's not working without regular expression

File: .github/workflows/ci.yaml (hunk 11 above)
Comment (Mixaster995) [reply .github/workflows/ci.yaml:184]: It's not working without regular expression

**Comment by Mixaster995** [comment]
@denis-tingaikin fixed all review comments, except regex in `go test -run ...`

**Comment by denis-tingaikin** [comment]
@Mixaster995 interdomain job has failed. Could you have a look?

File: .github/workflows/ci.yaml (hunk 5 above)
Comment (denis-tingaikin) [reply .github/workflows/ci.yaml:144]: @Mixaster995 still actual

File: .github/workflows/ci.yaml [hunk 16]
```diff
@@ -141,8 +141,12 @@ jobs:
         run: go get sigs.k8s.io/kind@v0.11.1
       - name: Create kind clusters
         run: |
-          for (( i = 0; i < 3; i++ )); do
-              kind create cluster --name "kind-${i}" --config cluster-config-interdomain.yaml --image="kindest/node:$KUBERNETES_VERSION"
+          IMAGE="kindest/node:$KUBERNETES_VERSION"
+          if [[ $KUBERNETES_VERSION=="" ]]; then
+              IMAGE="kindest/node:v1.22.1"
```
Comment (denis-tingaikin) [inline .github/workflows/ci.yaml:146]:
```suggestion
KUBERNETES_VERSION="1.22.1"
```

Comment (Mixaster995) [reply .github/workflows/ci.yaml]: ok, fixed

**Review by denis-tingaikin — APPROVED** [review]

**Pull request merged**